### PR TITLE
Mx 14324 sovereing refactor factories

### PR DESCRIPTION
**Pull request by raduchis**
## Reasoning behind the pull request
- refactored components that know of chainRunType to be created by factories
  
## Proposed changes
- added factories for creation of components that know of chainRunType

## Testing procedure
- tests pass if integration tests pass

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: epochStart/bootstrap/storageProcess.go [hunk 1]
```diff
@@ -236,26 +245,34 @@ func (sesb *storageEpochStartBootstrap) createStorageRequestHandler() (process.R
...
+		sovFactory, sovErr := requestHandlers.NewSovereignResolverRequestHandlerFactory(factory)
+		if sovErr != nil {
+			return nil, sovErr
+		}
+		sesb.resolverRequestFactory = sovFactory
 	default:
 		return nil, fmt.Errorf("%w type %v", errors.ErrUnimplementedChainRunType, sesb.chainRunType)
 	}
+
+	return sesb.resolverRequestFactory.CreateResolverRequestHandler(args)
```
Comment (sasurobert) [inline epochStart/bootstrap/storageProcess.go:275]: I think this is a little bit overkill - we can go with it. But it is so much more code. If in the end we can create components with less code - if that is the end goal, than it is fine.

File: process/smartContract/hooks/sovereignBlockChainHookFactory.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+		return nil, customErrors.ErrNilBlockChainHookFactory
+	}
+	return &sovereignBlockChainHookFactory{
+		blockChainHookFactory: blockChainHookFactory,
+	}, nil
+}
+
+// CreateBlockChainHook creates a blockchain hook based on the chain run type sovereign
+func (bhf *sovereignBlockChainHookFactory) CreateBlockChainHook(args ArgBlockChainHook) (process.BlockChainHookHandler, error) {
+	bh, _ := NewBlockChainHookImpl(args)
```
Comment (sasurobert) [inline process/smartContract/hooks/sovereignBlockChainHookFactory.go:26]: treat the error.

File: dataRetriever/requestHandlers/interface.go [hunk 3]
```diff
@@ -26,3 +33,19 @@ type HeaderRequester interface {
 	NonceRequester
 	EpochRequester
 }
+
+// ResolverRequestFactoryHandler defines the resolver requester factory handler
+type ResolverRequestFactoryHandler interface {
+	CreateResolverRequestHandler(resolverRequestArgs ResolverRequestArgs) (process.RequestHandler, error)
```
Comment (iulianpascalau) [inline dataRetriever/requestHandlers/interface.go:39]: Let's rename these a little bit: this function should be `CreateRequestHandler` as the returned type is `RequestHandler`. 
Suggestion to rename the interface to `RequestHandlerCreator` as it is shorter and better explains what it does.

File: dataRetriever/requestHandlers/interface.go [hunk 4]
```diff
@@ -26,3 +33,19 @@ type HeaderRequester interface {
...
+	IsInterfaceNil() bool
+}
+
+// ResolverRequestArgs holds all dependencies required by the process data factory to create components
+type ResolverRequestArgs struct {
```
Comment (iulianpascalau) [inline dataRetriever/requestHandlers/interface.go:44]: RequestHandlerArgs or ArgsRequestHandler

File: dataRetriever/requestHandlers/resolverRequestHandlerFactory_test.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	t.Parallel()
+
+	rrhf, _ := NewResolverRequestHandlerFactory()
+	require.False(t, rrhf.IsInterfaceNil())
+
+	rrhf = (*resolverRequestHandlerFactory)(nil)
+	require.True(t, rrhf.IsInterfaceNil())
+}
+
+func GetDefaultArgs() ResolverRequestArgs {
```
Comment (iulianpascalau) [inline dataRetriever/requestHandlers/resolverRequestHandlerFactory_test.go:41]: unexported function?

File: dataRetriever/requestHandlers/resolverRequestHandlerFactory.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package requestHandlers
```
Comment (iulianpascalau) [inline dataRetriever/requestHandlers/resolverRequestHandlerFactory.go:1]: optional: we might move these factories in dataRetriever/requestHandlers/factory sub-package

File: epochStart/bootstrap/interface.go [hunk 7]
```diff
@@ -60,3 +59,16 @@ type NodeTypeProviderHandler interface {
...
 }
+
+// EpochStartBootstrapperFactoryHandler defines the epoch start bootstrapper factory handler
+type EpochStartBootstrapperFactoryHandler interface {
+	CreateEpochStartBootstrapper(epochStartBootstrapArgs ArgsEpochStartBootstrap) (EpochStartBootstrapper, error)
+	IsInterfaceNil() bool
+}
+
+// EpochStartBootstrapper defines the epoch start bootstrap functionality
+type EpochStartBootstrapper interface {
```
Comment (iulianpascalau) [inline epochStart/bootstrap/interface.go:70]: :+1:

File: epochStart/bootstrap/interface.go [hunk 8]
```diff
@@ -60,3 +59,16 @@ type NodeTypeProviderHandler interface {
 	GetType() core.NodeType
 	IsInterfaceNil() bool
 }
+
+// EpochStartBootstrapperFactoryHandler defines the epoch start bootstrapper factory handler
+type EpochStartBootstrapperFactoryHandler interface {
```
Comment (iulianpascalau) [inline epochStart/bootstrap/interface.go:64]: suggestion to rename to `EpochStartBootstrapperCreator`

File: epochStart/bootstrap/interface.go [hunk 9]
```diff
@@ -60,3 +59,16 @@ type NodeTypeProviderHandler interface {
...
+	CreateEpochStartBootstrapper(epochStartBootstrapArgs ArgsEpochStartBootstrap) (EpochStartBootstrapper, error)
+	IsInterfaceNil() bool
+}
+
+// EpochStartBootstrapper defines the epoch start bootstrap functionality
+type EpochStartBootstrapper interface {
+	Bootstrap() (Parameters, error)
+	IsInterfaceNil() bool
```
Comment (iulianpascalau) [inline epochStart/bootstrap/interface.go:72]: Switch the order between Bootstrap and IsInterfaceNil ?

File: epochStart/bootstrap/epochStartBootstrapperFactory_test.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+	esbf, _ := NewEpochStartBootstrapperFactory()
+
+	require.False(t, esbf.IsInterfaceNil())
+
+	esbf = (*epochStartBootstrapperFactory)(nil)
+	require.True(t, esbf.IsInterfaceNil())
+}
+
+func GetDefaultArgs() ArgsEpochStartBootstrap {
```
Comment (iulianpascalau) [inline epochStart/bootstrap/epochStartBootstrapperFactory_test.go:59]: unexported function

File: epochStart/bootstrap/storageProcess.go [hunk 11]
```diff
@@ -33,10 +33,11 @@ import (
...
 type ArgsStorageEpochStartBootstrap struct {
 	ArgsEpochStartBootstrap
-	ImportDbConfig             config.ImportDbConfig
-	ChanGracefullyClose        chan endProcess.ArgEndProcess
-	TimeToWaitForRequestedData time.Duration
-	ChainRunType               common.ChainRunType
+	ImportDbConfig                       config.ImportDbConfig
+	ChanGracefullyClose                  chan endProcess.ArgEndProcess
+	TimeToWaitForRequestedData           time.Duration
+	ChainRunType                         common.ChainRunType
```
Comment (iulianpascalau) [inline epochStart/bootstrap/storageProcess.go:39]: why do we still have to provide the ChainRunType?

File: epochStart/bootstrap/storageProcess.go [hunk 12]
```diff
@@ -74,24 +81,26 @@ func NewStorageEpochStartBootstrap(args ArgsStorageEpochStartBootstrap) (*storag
...
+func createEpochStartBootstrapper(args ArgsStorageEpochStartBootstrap) (EpochStartBootstrapper, error) {
+	fact, err := NewEpochStartBootstrapperFactory()
 	if err != nil {
 		return nil, err
 	}
 
 	switch args.ChainRunType {
 	case common.ChainRunTypeRegular:
-		return esb, nil
+		args.EpochStartBootstrapperFactoryHandler = fact
```
Comment (iulianpascalau) [inline epochStart/bootstrap/storageProcess.go:92]: function that changes (locally, but changes) the arguments. Please refactor,

File: factory/bootstrap/bootstrapComponents.go [hunk 13]
```diff
@@ -44,16 +44,17 @@ type BootstrapComponentsFactoryArgs struct {
...
+	config                               config.Config
+	prefConfig                           config.Preferences
+	importDbConfig                       config.ImportDbConfig
+	flagsConfig                          config.ContextFlagsConfig
+	workingDir                           string
+	coreComponents                       factory.CoreComponentsHolder
+	cryptoComponents                     factory.CryptoComponentsHolder
+	networkComponents                    factory.NetworkComponentsHolder
+	statusCoreComponents                 factory.StatusCoreComponentsHolder
+	chainRunType                         common.ChainRunType
```
Comment (iulianpascalau) [inline factory/bootstrap/bootstrapComponents.go:56]: why do we still have the chainRunType here?

File: factory/consensus/consensusComponents.go [hunk 14]
```diff
@@ -57,22 +57,24 @@ type ConsensusComponentsFactoryArgs struct {
...
+	dataComponents                         factory.DataComponentsHolder
+	processComponents                      factory.ProcessComponentsHolder
+	stateComponents                        factory.StateComponentsHolder
+	statusComponents                       factory.StatusComponentsHolder
+	statusCoreComponents                   factory.StatusCoreComponentsHolder
+	scheduledProcessor                     consensus.ScheduledProcessor
+	isInImportMode                         bool
+	shouldDisableWatchdog                  bool
+	consensusModel                         consensus.ConsensusModel
+	chainRunType                           common.ChainRunType
```
Comment (iulianpascalau) [inline factory/consensus/consensusComponents.go:75]: why do we still need this?

File: process/smartContract/hooks/factory.go [hunk 15]
```diff
@@ -10,14 +10,22 @@ import (
...
 	case common.ChainRunTypeRegular:
-		return bh, err
+		args.BlockChainHookFactoryHandler = factory
 	case common.ChainRunTypeSovereign:
-		return NewSovereignBlockChainHook(bh)
+		sovereignFactory, sovErr := NewSovereignBlockChainHookFactory(factory)
+		if sovErr != nil {
+			return nil, sovErr
+		}
+		args.BlockChainHookFactoryHandler = sovereignFactory
```
Comment (iulianpascalau) [inline process/smartContract/hooks/factory.go:25]: function that alters the arguments. Please refactor. Why do we still need this function `CreateBlockChainHook`? We should use directly the instances of type  BlockChainHookFactoryHandler

File: process/smartContract/hooks/interface.go [hunk 16]
```diff
@@ -11,3 +14,9 @@ type BlockChainHookCounter interface {
 	GetCounterValues() map[string]uint64
 	IsInterfaceNil() bool
 }
+
+// BlockChainHookFactoryHandler defines the blockchain hook factory handler
+type BlockChainHookFactoryHandler interface {
```
Comment (iulianpascalau) [inline process/smartContract/hooks/interface.go:19]: BlockChainHookHandlerCreator ?

File: process/sync/storageBootstrap/interface.go [hunk 17]
```diff
@@ -17,3 +19,15 @@ type storageBootstrapperHandler interface {
 	getRootHash(hash []byte) []byte
 	IsInterfaceNil() bool
 }
+
+// ShardStorageBootstrapperFactoryHandler defines the operations supported by a shard storage bootstrapper factory
+type ShardStorageBootstrapperFactoryHandler interface {
+	CreateShardStorageBootstrapper(args ArgsShardStorageBootstrapper) (process.BootstrapperFromStorage, error)
```
Comment (iulianpascalau) [inline process/sync/storageBootstrap/interface.go:25]: CreateBootstrapperFromStorage ?

File: process/sync/storageBootstrap/interface.go [hunk 18]
```diff
@@ -17,3 +19,15 @@ type storageBootstrapperHandler interface {
 	getRootHash(hash []byte) []byte
 	IsInterfaceNil() bool
 }
+
+// ShardStorageBootstrapperFactoryHandler defines the operations supported by a shard storage bootstrapper factory
+type ShardStorageBootstrapperFactoryHandler interface {
```
Comment (iulianpascalau) [inline process/sync/storageBootstrap/interface.go:24]: BootstrapperFromStorageCreator ?

File: process/sync/storageBootstrap/interface.go [hunk 19]
```diff
@@ -17,3 +19,15 @@ type storageBootstrapperHandler interface {
...
+	CreateShardStorageBootstrapper(args ArgsShardStorageBootstrapper) (process.BootstrapperFromStorage, error)
+	IsInterfaceNil() bool
+}
+
+// ShardBootstrapFactoryHandler defines the operations supported by a shard bootstrap factory
+type ShardBootstrapFactoryHandler interface {
+	CreateShardBootstrapFactory(argsBaseBootstrapper sync.ArgShardBootstrapper) (process.Bootstrapper, error)
```
Comment (iulianpascalau) [inline process/sync/storageBootstrap/interface.go:31]: CreateBootstrapper ?

File: process/sync/storageBootstrap/interface.go [hunk 20]
```diff
@@ -17,3 +19,15 @@ type storageBootstrapperHandler interface {
...
 }
+
+// ShardStorageBootstrapperFactoryHandler defines the operations supported by a shard storage bootstrapper factory
+type ShardStorageBootstrapperFactoryHandler interface {
+	CreateShardStorageBootstrapper(args ArgsShardStorageBootstrapper) (process.BootstrapperFromStorage, error)
+	IsInterfaceNil() bool
+}
+
+// ShardBootstrapFactoryHandler defines the operations supported by a shard bootstrap factory
+type ShardBootstrapFactoryHandler interface {
```
Comment (iulianpascalau) [inline process/sync/storageBootstrap/interface.go:30]: BootstrapperCreator ?

File: testscommon/uint64ByteSliceConverterMock.go [hunk 21]
```diff
@@ -0,0 +1,22 @@
+package testscommon
+
+// Uint64ByteSliceConverterMock converts byte slice to/from uint64
+type Uint64ByteSliceConverterMock struct {
```
Comment (iulianpascalau) [inline testscommon/uint64ByteSliceConverterMock.go:4]: this is a stub, please rename the file & struct

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: dataRetriever/requestHandlers/interface.go (hunk 3 above)
Comment (raduchis) [reply dataRetriever/requestHandlers/interface.go:39]: good suggestion. Renamed

File: dataRetriever/requestHandlers/interface.go (hunk 4 above)
Comment (raduchis) [reply dataRetriever/requestHandlers/interface.go:44]: renamed

File: dataRetriever/requestHandlers/resolverRequestHandlerFactory.go (hunk 6 above)
Comment (raduchis) [reply dataRetriever/requestHandlers/resolverRequestHandlerFactory.go:1]: will move them in another PR

File: dataRetriever/requestHandlers/resolverRequestHandlerFactory_test.go (hunk 5 above)
Comment (raduchis) [reply dataRetriever/requestHandlers/resolverRequestHandlerFactory_test.go:41]: made unexported.

File: epochStart/bootstrap/epochStartBootstrapperFactory_test.go (hunk 10 above)
Comment (raduchis) [reply epochStart/bootstrap/epochStartBootstrapperFactory_test.go:59]: made unexported.

File: epochStart/bootstrap/interface.go (hunk 9 above)
Comment (raduchis) [reply epochStart/bootstrap/interface.go:72]: moved IsInterfaceNil to last position

File: epochStart/bootstrap/storageProcess.go (hunk 11 above)
Comment (raduchis) [reply epochStart/bootstrap/storageProcess.go:39]: Will remove chainRunType in one of the following PR-s. This PR is for extracting functionality in Creators/Factories without changing the current logic.

File: epochStart/bootstrap/interface.go (hunk 8 above)
Comment (raduchis) [reply epochStart/bootstrap/interface.go:64]: renamed

File: epochStart/bootstrap/storageProcess.go (hunk 12 above)
Comment (raduchis) [reply epochStart/bootstrap/storageProcess.go:92]: the switch will be removed in a next PR - added \\TODO  to remove switch

File: epochStart/bootstrap/storageProcess.go (hunk 1 above)
Comment (raduchis) [reply epochStart/bootstrap/storageProcess.go:275]: the entire switch will be removed once the creator/factories will be injected and just the last return will be needed

File: factory/consensus/consensusComponents.go (hunk 14 above)
Comment (raduchis) [reply factory/consensus/consensusComponents.go:75]: will be removed

File: process/smartContract/hooks/factory.go (hunk 15 above)
Comment (raduchis) [reply process/smartContract/hooks/factory.go:25]: the entire switch will be removed in a future PR. To keep everything working, just the factories were created and used.

File: process/smartContract/hooks/interface.go (hunk 16 above)
Comment (raduchis) [reply process/smartContract/hooks/interface.go:19]: renamed

File: process/smartContract/hooks/sovereignBlockChainHookFactory.go (hunk 2 above)
Comment (raduchis) [reply process/smartContract/hooks/sovereignBlockChainHookFactory.go:26]: fixed

File: process/sync/storageBootstrap/interface.go (hunk 18 above)
Comment (raduchis) [reply process/sync/storageBootstrap/interface.go:24]: done

**Review by sasurobert — APPROVED** [review]

File: epochStart/bootstrap/storageProcess.go (hunk 11 above)
Comment (iulianpascalau) [reply epochStart/bootstrap/storageProcess.go]: great

File: epochStart/bootstrap/storageProcess.go (hunk 12 above)
Comment (iulianpascalau) [reply epochStart/bootstrap/storageProcess.go]: great

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**